keps/sig-api-machinery/4006-transition-spdy-to-websockets/README.md

```diff
@@ -0,0 +1,1200 @@
+initial HTTP/1.1 request. By adding two headers (`Connection: Upgrade`, `Upgrade: SPDY/3.1`),
+the request can initiate the streaming upgrade. And when the response returns status
+`101 Switching Protocols` signalling success, the connection can then be kept open
+for subsequent streaming. An example of an upgraded HTTP Request/Response for `kubectl exec`
+could look like:
+
+**HTTP Request**
+```
+POST /api/v1/…/pods/nginx/exec?command=<CMD>... HTTP/1.1
+Connection: Upgrade
+Upgrade: SPDY/3.1
+X-Stream-Protocol-Version: v4.channel.k8s.io
+X-Stream-Protocol-Version: v3.channel.k8s.io
+X-Stream-Protocol-Version: v2.channel.k8s.io
+X-Stream-Protocol-Version: v1.channel.k8s.io
+```
+
+**HTTP Response**
+```
+HTTP/1.1 101 Switching Protocols
+Connection: Upgrade
+Upgrade: SPDY/3.1
+X-Stream-Protocol-Version: v4.channel.k8s.io
+```
+
+If the upgrade is successful, one of the requested subprotocol versions is chosen
+and returned in the response. In this instance, the chosen version of the subprotocol
+is: `v4.channel.k8s.io`.
+
+### Background: `RemoteCommand` Subprotocol
+
+![Remote Command Subprotocol](./remote-command-subprotocol.png)
+
+Once the connection is upgraded to a bi-directional streaming connection, the
+client and server can exchange data messages. These messages are interpreted with
+agreed upon standards which are called subprotocols. The three `kubectl` commands
+(`exec`, `attach`, and `cp`) communicate using the `RemoteCommand` subprotocol. Basically,
+this subprotocol provides command line functionality from the client to a running
+container in the cluster. By multiplexing `stdin`, `stdout`, `stderr`, and `tty`
+resizing over a streaming connection, this subprotocol supports clients executing
+and interacting with commands executed on a container in the cluster. An example of
+`kubectl exec` running the `date` command on an `nginx` pod/container is:
+
+```
+$ kubectl exec nginx -- date
+Tue May 16 03:34:04 PM PDT 2023
+```
+
+### Background: API Server and Kubelet `UpgradeAwareProxy`
+
+In order to route the data streamed between the client and the container, both the
+API Server and Kubelet must proxy these data messages. Both the API Server and the
+Kubelet provide this functionality with the `UpgradeAwareProxy`, which is a reverse
+proxy that knows how to deal with the connection upgrade handshake.
+
+### Proposal: `kubectl` WebSocket Executor and Fallback Executor
+
+This enhancement proposes adding a `WebSocketExecutor` to `kubectl`, implementing
+the WebSocket client using a new subprotol version (`v5.channel.k8s.io`). Additionally,
+we propose creating a `FallbackExecutor` to address client/server version skew. The
+`FallbackExecutor`  first attempts to upgrade the connection to WebSockets
+version five (`v5.channel.k8s.io`) with the `WebSocketExecutor`, then falls back
+to the legacy `SPDYExecutor` version four (`v4.channel.k8s.io`), if the upgrade is
```

Comment:
Why it certainly falls back to version four (`v4.channel.k8s.io`)?. Since we are using legacy SPDY executor and if very old API server uses `v3.channel.k8s.io`, SPDY executor can fall back to version three?

Reply:
there is only n-3 versions supported, we should focus only on the supported versions

Reply:
So the current SPDY implementation asks for four versions, with `v4.channel.k8s.io` being the highest priority (`v3` next, etc.). The request headers looks like
```
I0518 16:16:00.390959  633715 round_trippers.go:469] Request Headers:
I0518 16:16:00.390963  633715 round_trippers.go:473]     X-Stream-Protocol-Version: v4.channel.k8s.io
I0518 16:16:00.390967  633715 round_trippers.go:473]     X-Stream-Protocol-Version: v3.channel.k8s.io
I0518 16:16:00.390969  633715 round_trippers.go:473]     X-Stream-Protocol-Version: v2.channel.k8s.io
I0518 16:16:00.390972  633715 round_trippers.go:473]     X-Stream-Protocol-Version: channel.k8s.io
```
Since it has been so many releases since `v4` has been implemented, this is almost always the version selected in the handshake.

I do not think we should change anything about the current SPDY implementation. In other words, I believe it would be fine to keep the request for these four versions (in which it is almost certain the `v4` version is chosen).